Improve filtering on table pages

DL6ER opened the pull request:
**By submitting this pull request, I confirm the following:** 

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
**What does this PR aim to accomplish?:**

Improve filtering on table pages. This enables a whole set of new features, see https://discourse.pi-hole.net/t/explicit-search-in-pi-hole-admin-console/28893/16

**How does this PR accomplish the above?:**

Change on-click actions on query types, domain and client from searching to filtering

**What documentation changes (if any) are needed to support this PR?:**

None

DL6ER committed:
Add exact searching using modified regex for query type, domain, and client.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Initialize regex filter mode and clear search field (if set previously)

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add [Ctrl] key action to combine filters.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Show filtering details on clear button.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add missing space in front of and on the dynamically generated button text.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add comments and simplify code

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add comment that holding down [Ctrl] can do something magical

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Improve wording

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add meta key modifier and highlight filtering columns

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Reduce code duplication

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Remove pointer cursor when auto-filtering is disabled.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Replace / by "or".

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Store filtering columns in array instead of in a string. This makes adding a column multiple times impossible.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Implement selective undoing using the Shift key.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add filtering on status column.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add reply type filtering.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Apply filtering events only on Ctrl/Command/Shift + Click.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Simply reply type code.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Use classes for highlighting to ensure the alternate row colors are not lost when removing a column from the selection.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Fix wording

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Merge branch 'devel' into tweak/exact_searching

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Minor style changes for the dark theme.

Signed-off-by: DL6ER <dl6er@dl6er.de>

XhmikosR commented on scripts/pi-hole/js/queries.js:
```diff
@@ -496,28 +432,108 @@ $(function () {
   });
-});
+
+  button.text(text);
```
Did you check if this looks good with long content? Maybe we shouldn't change the button's text, and do it with a span above it or something?

yubiuser replied:
https://discourse.pi-hole.net/t/explicit-search-in-pi-hole-admin-console/28893/57?u=yubiuser

PromoFaux replied:
I think probably a "Clear current filter" would suffice. That is what it is doing, after all.

XhmikosR commented on scripts/pi-hole/js/utils.js:
```diff
@@ -198,7 +198,14 @@ function stateLoadCallback(itemName) {
   data = JSON.parse(data);
+
+  // Clear possible filtering settings
+  data.columns.forEach(function (value, index) {
```
Needs a confirmation that other places that use `stateLoadCallback()` are OK.

DL6ER replied:
Yes, we are not using column filtering anywhere. However, even if we did, we'd like to have this cleared on page revisiting to prevent hidden filtering.

XhmikosR commented on queries.php:
```diff
@@ -138,11 +138,11 @@
-                <label for="autofilter">Apply filtering on click on Type, Domain, and Clients</label>
-            </div>
-            <button type="button" id="resetButton" class="btn btn-default btn-sm text-red hidden">Clear Filters</button>
+            <label>Filtering options:</label>
```
Shouldn't be a `label` since there's no form AFAICT.

XhmikosR commented on scripts/pi-hole/js/queries.js:
```diff
@@ -496,28 +432,108 @@ $(function () {
+      regex = "^" + value + "$";
+
+      // Add background color
+      tableApi.$("td:eq(" + index + ")").addClass("highlight");
```
The highlight color is too light...

DL6ER replied:
This is personal taste. It is the same color we use also in the domain list when highlighting a regex match. The color was discussed in the Discourse thread and I think we agreed that it is clear enough. You also don't want to make reading the cell content difficult.

![Screenshot from 2020-06-03 19-46-30](https://user-images.githubusercontent.com/16748619/83670184-1196cf00-a5d3-11ea-8573-b5ad9e5a9870.png)

yubiuser replied:
Agree. I think the color is rather obvious (for non color-blind at least)

yubiuser replied:
Just switched to the branch and saw that the filter color for me is indeed quite aggressiv. Is there a reason why I have this strong yellow and you the sand colored? I remember that I used to have the sand color also when I tried the branch last time (27.04.)

![Bildschirmfoto zu 2020-06-03 22-30-24](https://user-images.githubusercontent.com/26622301/83687239-c76e1780-a5eb-11ea-9bae-bf7d91addc9f.png)

XhmikosR commented:
Also, there's no indication that there is an interaction... The note at the bottom is nice, but the whole thing is not accessible at all.

PromoFaux commented:
I like it on the main, just a couple of thoughts:

Maybe it would be more user friendly to have the filtering hints at the top of the page?

Maybe a duplicate of the clear button at the top of the table?

DL6ER commented:
> but the whole thing is not accessible at all.

How would you like to see this improved? Note that this is also by no means an essential function, it is more a "nice to have" for power users. I think the hint should be enough, moving it to the top seems distracting as it causes the actual content to go down the page. This is not worth making the hint a bit more obvious. We will highlight this in the v5.1 changelog/blog post so more people get to know it. As said, I think only a smaller fraction of users actually use it, however, for those the feature will be very useful.

XhmikosR commented:
This isn't a matter if people know something. It's a matter of accessibility which we are really lacking of.

There should be at least a visual indication for normal mouse users. Not sure how to make this proper for the other users, since that accessibility-wise buttons or `a` elements with `role="button"` are better IIRC.

DL6ER commented:
I'm open for discussions. It is intentional that there is no cursor pointer as clicking without holding <kbd>Ctrl</kbd> isn't doing anything. Maybe we can do something with a tooltip here. There's a lot of discussion on usability on the liked Discourse thread. Maybe @yubiuser recalls better than me what we discussed. I will check tomorrow, otherwise.

yubiuser commented:
We discussed mainly how to select/deselect a filter. We came up with the solution now proposed: normal click does nothing, click+Ctrl adds column to filter, click+shift removes from filter. This seemed to be the easiest solution and we got rid of the checkbox "Apply filtering on click...." and kept the ability to filter and search.

Open points where the text on the bottom and the text on the clear button... and of course the extent to the long-term database query page ;-)

I would keep the text on the bottom, but maybe add a tooltip: if column is nit already filterd "Click+Ctrl to filter" and if column is already filtered "Click+Shift to remove filter"

DL6ER committed:
Simplify button by removing the description of what we are filtering.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add tooltip and click cursor to enhance accessibility of the click filtering feature.

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Reduce agressiveness of highlight color.

Signed-off-by: DL6ER <dl6er@dl6er.de>

XhmikosR commented on scripts/pi-hole/js/queries.js:
```diff
@@ -388,33 +389,83 @@ $(function () {
+            $(this).css("cursor", "pointer").attr("title", tooltipText(1, this.textContent));
+          },
+          function () {
+            $(this).css("cursor", "auto");
```
Please don't do this. We already have CSS classes to use. It's always more performant to add/remove CSS classes that this.

Also, do note that this does not fix anything accessibility related. It's just a visual indication only.

DL6ER replied:
So what is the right way to do this?

XhmikosR replied:
`addClass('pointer')`/`removeClass('pointer')`

There is a reason we don't have such utils upstream BTW. Because adding a cursor does not improve accessibility, it's only a visual indication, which on non-pointer devices does not help at all. :)

DL6ER replied:
Thanks. I don't think the web interface is meant or designed for non-pointer devices. I will check, nonetheless, if we can have the same effect with a stepping through using <kbd>Tab</kbd> and then maybe <kbd>Ctrl</kbd>+<kbd>Space</kbd>.

DL6ER commented:
I addressed the raised review points:

- The button text is not short and static. Filtering is obvious enough from the highlighted columns
- I fixed the highlight color. They got lost in the adjustments for the light/dark themes. The highlighting colors of both themes are now set up correctly (it is intentional to change them theme-wide).
- I added both click pointers as well as dynamic tooltips:
    ![Screenshot from 2020-06-04 10-06-18](https://user-images.githubusercontent.com/16748619/83732564-89551000-a64c-11ea-8ccd-ad7c0ac09d2f.png)
    ![Screenshot from 2020-06-04 10-11-56](https://user-images.githubusercontent.com/16748619/83732571-8b1ed380-a64c-11ea-8500-64086177895f.png)
    (the cursor isn't shown in the screenshots, however, it is of pointer type.

XhmikosR commented on scripts/pi-hole/js/queries.js:
```diff
@@ -306,49 +306,17 @@ $(function () {
-            replytext = "? (" + parseInt(data[6]) + ")";
-        }
+      var replytext,
+        replyid = parseInt(data[5]);
```
Can you try to specify the radix from now on? I do have #1374 but not sure when that will be tackled.

DL6ER replied:
Sure, I will look at #1374 soon as well.

DL6ER replied:
I removed `parseInt()` here as we don't need it.

DL6ER committed:
Use classes instead of manipulating CSS directly for getting the pointer.

Signed-off-by: DL6ER <dl6er@dl6er.de>

yubiuser commented:
Thanks I like it a lot.

DL6ER committed:
Remove parseInt() from values guaranteed to be int from the API.

Signed-off-by: DL6ER <dl6er@dl6er.de>

yubiuser commented:
Initially filtering on status and reply was not possible due to technical reasons. Later you added the ability.
Does this now also enable the ability to search these columns? If not possible maybe add "Type/Domain/Client" as light grey text in the search input field so users know that it is limited to those columns.

DL6ER committed:
Add placeholder to the search field to highlight what can be searched for

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
> Initially filtering on status and reply was not possible due to technical reasons. Later you added the ability.

Where did I add this? :thinking: 
I don't think it is needed, there is a rather limited set of possibilities for these fields and filtering-on-click should be totally sufficient. Still, the placeholder is a good idea.

![Screenshot from 2020-06-04 10-52-40](https://user-images.githubusercontent.com/16748619/83736222-99232300-a651-11ea-928d-cb47f9506f74.png)

yubiuser commented:
> Where did I add this?

https://discourse.pi-hole.net/t/explicit-search-in-pi-hole-admin-console/28893/36?u=yubiuser
https://discourse.pi-hole.net/t/explicit-search-in-pi-hole-admin-console/28893/39?u=yubiuser

DL6ER commented:
Ah, I see. This was a misunderstanding. Yes, filtering on click is possible, but not through the search. The new placeholder should clarify this. Thanks.

DL6ER committed:
Merge branch 'devel' into tweak/exact_searching

DL6ER commented:
@PromoFaux Kindly requesting re-approval.

XhmikosR commented on style/themes/default-dark.css:
```diff
@@ -385,7 +385,7 @@ pre {
 }
 td.highlight {
-  background-color: yellow;
+  background-color: #ffcc0050;
```
Please don't use this notation. Just use `rgba`

XhmikosR commented on scripts/pi-hole/js/queries.js:
```diff
@@ -8,6 +8,22 @@
 /* global moment:false, utils:false */
 
 var tableApi;
+var tableFilters = [];
```
Do the newly added variables really need to be global?

DL6ER replied:
Yes, it needs to be available all the time.

DL6ER committed:
CSS: Use rgba() instead of hex-color.

Signed-off-by: DL6ER <dl6er@dl6er.de>

PromoFaux reviewed (approved):
Functionally speaking, this works well.

reinstej-sci commented:
Were consequences of this change on mobile/tablet browsers considered? This has broken the ability to click to filter on mobile.
